File: Data Structures & Algorithms/permutations/submission-1.js
class Solution {
    res = []
    nums = []
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    permute(nums) {
        this.nums = nums
        let bstate = new Array(this.nums.length).fill(false)
        this.foo(0, [], bstate)
        return this.res
    }

    foo(idx, state, bstate) {
        if(idx === this.nums.length) {
            this.res.push([...state])
            return
        }

        for(let i = 0; i < this.nums.length; i++) {
            let n = this.nums[i]

            if(!bstate[i]) {
                state.push(n)
                bstate[i] = true
                this.foo(idx + 1, state, bstate)
                state.pop()
                bstate[i] = false
            }
        }
    }
}

File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:
      vector<int> nums;
  vector<vector<int>> res;

  vector<vector<int>> subsets(vector<int>& nums) {
    this->nums = nums;
    vector<int> state;

    this->dfs(state, 0);

    return this->res;
  }

  void dfs(vector<int>& state, int i) {
    if (i == this->nums.size()) {
      auto copy = vector(state);
      this->res.push_back(copy);
      return;
    }

    state.push_back(this->nums[i]);
    dfs(state, i+1);
    state.pop_back();
    dfs(state, i+1);
  }

};
